_posts/2019-05-06-live-view-with-presence.md

```diff
@@ -0,0 +1,587 @@
+
+In order to respond to the event we'll define a `handle_info/2` function in our live view that will match the `"presence_diff"` event:
+
+```elixir
+# lib/phat_web/live/chat_live_view.ex
+
+def handle_info(%{event: "presence_diff"}, socket = %{assigns: %{chat: chat}}) do
+
+end
+```
+
+This function has two responsibilities:
+
+* Get the list of present users for the given chat room topic from the Presence data store
+* Update the LiveView socket's state to reflect this list of users
+
+```elixir
+def handle_info(%{event: "presence_diff", payload: _payload}, socket = %{assigns: %{chat: chat}}) do
+  users =   
+    Presence.list(topic(chat.id))
+    |> Enum.map(fn {_user_id, data} ->
+      data[:metas]
+      |> List.first()
+    end)
+
+  {:noreply,
+   assign(socket,
+     users: users
+   )}
+end
+```
+
+First, we use the `Presence.list/1` function to get the collection of present users under the given topic. This will return the following data structure:
+
+```elixir
+%{
+  "1" => %{
+    metas: [
+      %{
+        email: "sophie@email.com",
+        first_name: "Sophie",
+        phx_ref: "TNV4PzRfyhw="
+        user_id: 1
+      }
+  },
+  "2" => %{
+    metas: [
+      %{
+        email: "beini@email.com",
+        first_name: "Beini",
+        phx_ref: "ZZ30QuoI/8s="
+        user_id: 1
+      }
+  }
+  ...
+}
+```
+
+The Presence behavior handles the diffs of join and leave events for us. So, as long as we call `Presence.track/4`, the Presence process will update its own state, such that when we next call `Presence.list/1`, we are retrieving the updated list of users.
+
+Once we fetch this list, we iterate over it to collect a list of the individual `:metas` payloads that describe each user. The resulting list will look like this:
+
+```elixir
+[
+  %{
+    email: "sophie@email.com",
+    first_name: "Sophie",
+    phx_ref: "TNV4PzRfyhw="
+    user_id: 1
+  },
+  "2" => %{
+    metas: [
+      %{
+        email: "beini@email.com",
+        first_name: "Beini",
+        phx_ref: "ZZ30QuoI/8s="
+        user_id: 1
+      }
+  }
+]
+```
+
+We enact this transformation so that we have a simple, easy-to-use data structure to interact with in the template when we want to list present user names.
+
+Lastly, we update the LiveView socket's state by adding a key of `:users` pointing to a value of our user list:
+
+```elixir
+{:noreply, assign(socket, users: users)}
+```
+
+Now we can access the user list via the `@users` assignment in our template to list out the names of the users present in the chatroom:
+
+```elixir
+# lib/phat_web/templates/chat/show.html.leex
+
+<h3>Members</h3>
+<%= for user <- @users do %>
+  <p>
+    <%= user.first_name %>
+  </p>
+<% end %>
+```
+
```

Comment:
Add a note here that now the application will break now, as it does not know nothing about `@users`. It will come back to working state after the next step however, where you handle them in `mount/2`.